File: banker.c
//********************************************************************
//
// Name: Anjan Roy
// Class: CSC 360 - Operating Systems 
// Programming Project 04: Implementatin of Banker's Algorithm
// July 7th, 2024
// Instructor: Dr. Siming Liu
//
//********************************************************************
#include <stdio.h>
#include <stdlib.h>

// Function: readFile
// Description: Reads the input file and initializes the allocation, max, available, and request matrices.
// Value Parameter Data Dictionary:
// - filename: char pointer to the name of the input file.
// Reference Parameter Data Dictionary:
// - n: integer reference to store the number of processes.
// - m: integer reference to store the number of resource types.
// - allocation: double pointer to store the allocation matrix.
// - max: double pointer to store the max matrix.
// - available: pointer to store the available vector.
// - requestProcess: integer reference to store the process making the request.
// - request: pointer to store the request vector.
void readFile(char *filename);

// Function: printMatrices
// Description: Prints the allocation and max matrices.
// Value Parameter Data Dictionary: None
// Reference Parameter Data Dictionary:
// - allocation: double pointer to the allocation matrix.
// - max: double pointer to the max matrix.
// - n: integer reference to the number of processes.
// - m: integer reference to the number of resource types.
void printMatrices();

// Function: calculateNeedMatrix
// Description: Calculates the need matrix using the allocation and max matrices.
// Value Parameter Data Dictionary: None
// Reference Parameter Data Dictionary:
// - need: double pointer to the need matrix.
// - allocation: double pointer to the allocation matrix.
// - max: double pointer to the max matrix.
// - n: integer reference to the number of processes.
// - m: integer reference to the number of resource types.
void calculateNeedMatrix();

// Function: isSafe
// Description: Checks if the system is in a safe state.
// Value Parameter Data Dictionary: None
// Reference Parameter Data Dictionary:
// - n: integer reference to the number of processes.
// - m: integer reference to the number of resource types.
// - allocation: double pointer to the allocation matrix.
// - need: double pointer to the need matrix.
// - available: pointer to the available vector.
// Local Variable Data Dictionary:
// - work: pointer to store a copy of the available vector.
// - finish: pointer to track which processes can finish.
// - safeSequence: array to store the sequence of safe processes.
int isSafe();

// Function: canRequestBeGranted
// Description: Checks if a resource request can be granted and updates matrices if the request is safe.
// Value Parameter Data Dictionary:
// - process: integer value of the requesting process.
// - request: pointer to the request vector.
// Reference Parameter Data Dictionary:
// - n: integer reference to the number of processes.
// - m: integer reference to the number of resource types.
// - allocation: double pointer to the allocation matrix.
// - need: double pointer to the need matrix.
// - available: pointer to the available vector.
// Local Variable Data Dictionary:
// - safe: integer to store if the system is safe after granting the request.
int canRequestBeGranted(int process, int request[]);

// Function: printVector
// Description: Prints a vector with a title.
// Value Parameter Data Dictionary:
// - title: char pointer to the title of the vector.
// - vector: pointer to the vector.
// - length: integer value of the length of the vector.
// Reference Parameter Data Dictionary: None
void printVector(char *title, int *vector, int length);

// Function: printMatrix
// Description: Prints a matrix with a title.
// Value Parameter Data Dictionary:
// - title: char pointer to the title of the matrix.
// - matrix: double pointer to the matrix.
// - rows: integer value of the number of rows.
// - cols: integer value of the number of columns.
// Reference Parameter Data Dictionary: None
void printMatrix(char *title, int **matrix, int rows, int cols);

int n, m;
int **allocation, **max, **need;
int *available, *request;
int requestProcess;

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <input file>\n", argv[0]);
        exit(EXIT_FAILURE);
    }

    readFile(argv[1]);
    printMatrices();
    calculateNeedMatrix();
    printMatrix("The Need Matrix is...", need, n, m);
    printVector("The Available Vector is...", available, m);

    if (isSafe()) {
        printf("THE SYSTEM IS IN A SAFE STATE!\n");
    } else {
        printf("THE SYSTEM IS NOT IN A SAFE STATE!\n");
    }

    printf("The Request Vector is...\n");
    for (int j = 0; j < m; j++) {
        printf("%d ", request[j]);
    }
    printf("\n");

    if (canRequestBeGranted(requestProcess, request)) {
        printf("THE REQUEST CAN BE GRANTED!\n");
    } else {
        printf("THE REQUEST CANNOT BE GRANTED!\n");
    }

    return 0;
}

void readFile(char *filename) {
    FILE *file = fopen(filename, "r");
    if (!file) {
        perror("Error opening file");
        exit(EXIT_FAILURE);
    }

    fscanf(file, "%d", &n);
    fscanf(file, "%d", &m);

    // Allocate memory for matrices and vectors
    allocation = (int **)malloc(n * sizeof(int *));
    max = (int **)malloc(n * sizeof(int *));
    need = (int **)malloc(n * sizeof(int *));
    for (int i = 0; i < n; i++) {
        allocation[i] = (int *)malloc(m * sizeof(int));
        max[i] = (int *)malloc(m * sizeof(int));
        need[i] = (int *)malloc(m * sizeof(int));
    }
    available = (int *)malloc(m * sizeof(int));
    request = (int *)malloc(m * sizeof(int));

    // Read allocation matrix
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            fscanf(file, "%d", &allocation[i][j]);
        }
    }

    // Read max matrix
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            fscanf(file, "%d", &max[i][j]);
        }
    }

    // Read available vector
    for (int j = 0; j < m; j++) {
        fscanf(file, "%d", &available[j]);
    }

    // Read request vector
    fscanf(file, "%d:", &requestProcess);
    for (int j = 0; j < m; j++) {
        fscanf(file, "%d", &request[j]);
    }

    fclose(file);
}

void printMatrices() {
    printMatrix("The Allocation Matrix is...", allocation, n, m);
    printMatrix("The Max Matrix is...", max, n, m);
}

void calculateNeedMatrix() {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            need[i][j] = max[i][j] - allocation[i][j];
        }
    }
}

int isSafe() {
    int *work = (int *)malloc(m * sizeof(int));
    int *finish = (int *)malloc(n * sizeof(int));

    for (int i = 0; i < m; i++) {
        work[i] = available[i];
    }
    for (int i = 0; i < n; i++) {
        finish[i] = 0;
    }

    int safeSequence[n];
    int count = 0;

    while (count < n) {
        int found = 0;
        for (int i = 0; i < n; i++) {
            if (finish[i] == 0) {
                int j;
                for (j = 0; j < m; j++) {
                    if (need[i][j] > work[j]) {
                        break;
                    }
                }
                if (j == m) {
                    for (int k = 0; k < m; k++) {
                        work[k] += allocation[i][k];
                    }
                    safeSequence[count++] = i;
                    finish[i] = 1;
                    found = 1;
                }
            }
        }
        if (found == 0) {
            free(work);
            free(finish);
            return 0; // System is not in a safe state
        }
    }

    free(work);
    free(finish);
    return 1; // System is in a safe state
}

int canRequestBeGranted(int process, int request[]) {
    for (int j = 0; j < m; j++) {
        if (request[j] > need[process][j]) {
            return 0; // Request exceeds maximum claim
        }
    }
    for (int j = 0; j < m; j++) {
        if (request[j] > available[j]) {
            return 0; // Not enough resources available
        }
    }

    // Temporarily allocate resources
    for (int j = 0; j < m; j++) {
        available[j] -= request[j];
        allocation[process][j] += request[j];
        need[process][j] -= request[j];
    }

    int safe = isSafe();

    // Revert allocation
    for (int j = 0; j < m; j++) {
        available[j] += request[j];
        allocation[process][j] -= request[j];
        need[process][j] += request[j];
    }

    return safe;
}

void printVector(char *title, int *vector, int length) {
    printf("%s\n", title);
    for (int i = 0; i < length; i++) {
        printf("%d ", vector[i]);
    }
    printf("\n");
}

void printMatrix(char *title, int **matrix, int rows, int cols) {
    printf("%s\n", title);
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            printf("%d ", matrix[i][j]);
        }
        printf("\n");
    }
}

File: banker.py
'''

    Name: Anjan Roy
    Class: CSC 360 - Operating Systems
    Programming Project 04: Implementation of Banker's Algorithm
    Instructor: Dr. Siming Liu

'''
import sys

def read_file(filename):
    """
    Function: read_file
    Description: Reads the input file and initializes the allocation, max, available, and request matrices.
    Value Parameter Data Dictionary:
    - filename: str, name of the input file.
    Reference Parameter Data Dictionary:
    - n: int, number of processes.
    - m: int, number of resource types.
    - allocation: list of lists, allocation matrix.
    - max_matrix: list of lists, max matrix.
    - available: list, available vector.
    - request_process: int, process making the request.
    - request: list, request vector.
    """
    with open(filename, 'r') as file:
        n = int(file.readline())
        m = int(file.readline())

        allocation = []
        for _ in range(n):
            allocation.append(list(map(int, file.readline().split())))

        max_matrix = []
        for _ in range(n):
            max_matrix.append(list(map(int, file.readline().split())))

        available = list(map(int, file.readline().split()))

        request_line = file.readline().split(':')
        request_process = int(request_line[0])
        request = list(map(int, request_line[1].split()))

    return n, m, allocation, max_matrix, available, request_process, request

def print_matrix(title, matrix):
    """
    Function: print_matrix
    Description: Prints a matrix with a title.
    Value Parameter Data Dictionary:
    - title: str, title of the matrix.
    - matrix: list of lists, matrix to print.
    Reference Parameter Data Dictionary: None
    """
    print(title)
    for row in matrix:
        print(' '.join(map(str, row)))

def print_vector(title, vector):
    """
    Function: print_vector
    Description: Prints a vector with a title.
    Value Parameter Data Dictionary:
    - title: str, title of the vector.
    - vector: list, vector to print.
    Reference Parameter Data Dictionary: None
    """
    print(title)
    print(' '.join(map(str, vector)))

def calculate_need_matrix(n, m, max_matrix, allocation):
    """
    Function: calculate_need_matrix
    Description: Calculates the need matrix using the allocation and max matrices.
    Value Parameter Data Dictionary: None
    Reference Parameter Data Dictionary:
    - n: int, number of processes.
    - m: int, number of resource types.
    - max_matrix: list of lists, max matrix.
    - allocation: list of lists, allocation matrix.
    - need: list of lists, need matrix.
    """
    need = [[0] * m for _ in range(n)]
    for i in range(n):
        for j in range(m):
            need[i][j] = max_matrix[i][j] - allocation[i][j]
    return need

def is_safe(n, m, allocation, need, available):
    """
    Function: is_safe
    Description: Checks if the system is in a safe state.
    Value Parameter Data Dictionary: None
    Reference Parameter Data Dictionary:
    - n: int, number of processes.
    - m: int, number of resource types.
    - allocation: list of lists, allocation matrix.
    - need: list of lists, need matrix.
    - available: list, available vector.
    Local Variable Data Dictionary:
    - work: list, copy of the available vector.
    - finish: list, track which processes can finish.
    - safe_sequence: list, sequence of safe processes.
    """
    work = available[:]
    finish = [False] * n
    safe_sequence = []

    while len(safe_sequence) < n:
        found = False
        for i in range(n):
            if not finish[i]:
                if all(need[i][j] <= work[j] for j in range(m)):
                    for j in range(m):
                        work[j] += allocation[i][j]
                    safe_sequence.append(i)
                    finish[i] = True
                    found = True
        if not found:
            return False
    return True

def can_request_be_granted(process, request, n, m, allocation, need, available):
    """
    Function: can_request_be_granted
    Description: Checks if a resource request can be granted and updates matrices if the request is safe.
    Value Parameter Data Dictionary:
    - process: int, process making the request.
    - request: list, request vector.
    Reference Parameter Data Dictionary:
    - n: int, number of processes.
    - m: int, number of resource types.
    - allocation: list of lists, allocation matrix.
    - need: list of lists, need matrix.
    - available: list, available vector.
    Local Variable Data Dictionary:
    - safe: bool, if the system is safe after granting the request.
    """
    if any(request[j] > need[process][j] for j in range(m)):
        return False
    if any(request[j] > available[j] for j in range(m)):
        return False

    for j in range(m):
        available[j] -= request[j]
        allocation[process][j] += request[j]
        need[process][j] -= request[j]

    safe = is_safe(n, m, allocation, need, available)

    for j in range(m):
        available[j] += request[j]
        allocation[process][j] -= request[j]
        need[process][j] += request[j]

    return safe

def main():
    """
    Function: main
    Description: Main function to execute the Banker's algorithm.
    Value Parameter Data Dictionary: None
    Reference Parameter Data Dictionary: None
    """
    if len(sys.argv) != 2:
        print("Usage: python banker.py <input file>")
        sys.exit(1)

    filename = sys.argv[1]
    n, m, allocation, max_matrix, available, request_process, request = read_file(filename)

    print(f"There are {n} processes in the system.")
    print(f"There are {m} resource types.")
    print_matrix("The Allocation Matrix is...", allocation)
    print_matrix("The Max Matrix is...", max_matrix)

    need = calculate_need_matrix(n, m, max_matrix, allocation)
    print_matrix("The Need Matrix is...", need)
    print_vector("The Available Vector is...", available)

    if is_safe(n, m, allocation, need, available):
        print("THE SYSTEM IS IN A SAFE STATE!")
    else:
        print("THE SYSTEM IS NOT IN A SAFE STATE!")

    print("The Request Vector is...")
    print(f"{request_process}:{' '.join(map(str, request))}")

    if can_request_be_granted(request_process, request, n, m, allocation, need, available):
        print("THE REQUEST CAN BE GRANTED!")
    else:
        print("THE REQUEST CANNOT BE GRANTED!")

if __name__ == "__main__":
    main()
